amrex@21.03: segfault during compile w/ gcc@8.1.0 on RHEL7

`amrex@21.03` fails to build using:
* `spack@develop` (195341113e7da85611553d1ea9bd907ccabeff64 from `Mon Mar 15 21:38:35 2021 +0100`)
* GCC 8.1.0 (built with Spack)
* Docker images:
  *  `ecpe4s/rhel7-runner-x86_64:2021-03-01`
  *  `ecpe4s/rhel7-runner-ppc64le:2021-03-01`


Concrete spec: [amrex-fze7ei.spec.yaml.txt](https://github.com/spack/spack/files/6094290/amrex-fze7ei.spec.yaml.txt)
Configure args: [amrex-spack-configure-args.txt](https://github.com/spack/spack/files/6094300/amrex-spack-configure-args.txt)
Build log: [amrex-spack-build-out.txt](https://github.com/spack/spack/files/6145088/amrex-spack-build-out.txt)


```
$> spack install -f ./amrex-fze7ei.spec.yaml
...
==> Installing amrex-21.03-fze7eilhl2hqne5r5u7g2qbxtex5uc2x
==> Fetching https://spack-llnl-mirror.s3-us-west-2.amazonaws.com/_source-cache/archive/63/6307bf75c80c2076bf5bd1cff4d12483280a32b5175fe117f32eed9c89cd1ac5.tar.gz
######################################################################## 100.0%
==> No patches needed for amrex
==> amrex: Executing phase: 'cmake'
==> amrex: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j32'

3 errors found in build log:
     271    /tmp/root/spack-stage/spack-stage-amrex-21.03-fze7eilhl2hqne5r5u7g2qbxtex5uc2x/spack-src/Src/LinearSolvers/MLMG/AMReX_MLNodeLaplacian.cpp:2061:21: internal compiler
             error: Segmentation fault
     272                         });
     273                         ^
     274    Please submit a full bug report,
     275    with preprocessed source if appropriate.
     276    See <https://github.com/spack/spack/issues> for instructions.
  >> 277    make[2]: *** [Src/CMakeFiles/amrex.dir/LinearSolvers/MLMG/AMReX_MLNodeLaplacian.cpp.o] Error 1
     278    make[2]: *** Waiting for unfinished jobs....
     279    make[2]: Leaving directory `/tmp/root/spack-stage/spack-stage-amrex-21.03-fze7eilhl2hqne5r5u7g2qbxtex5uc2x/spack-build-fze7eil'
  >> 280    make[1]: *** [Src/CMakeFiles/amrex.dir/all] Error 2
     281    make[1]: Leaving directory `/tmp/root/spack-stage/spack-stage-amrex-21.03-fze7eilhl2hqne5r5u7g2qbxtex5uc2x/spack-build-fze7eil'
  >> 282    make: *** [all] Error 2

See build log for details:
  /tmp/root/spack-stage/spack-stage-amrex-21.03-fze7eilhl2hqne5r5u7g2qbxtex5uc2x/spack-build-out.txt
```

Backtrace
```
/tmp/root/spack-stage/spack-stage-amrex-21.03-fze7eilhl2hqne5r5u7g2qbxtex5uc2x/spack-src/Src/LinearSolvers/MLMG/AMReX_MLNodeLaplacian.cpp: In lambda function:
/tmp/root/spack-stage/spack-stage-amrex-21.03-fze7eilhl2hqne5r5u7g2qbxtex5uc2x/spack-src/Src/LinearSolvers/MLMG/AMReX_MLNodeLaplacian.cpp:2061:21: internal compiler error: Segmentation fault
                     });
                     ^
0xb34b5f crash_signal
	/tmp/root/spack-stage/spack-stage-gcc-8.1.0-fde6qts62r4baorvcrpkjxc7c6vzmn3w/spack-src/gcc/toplev.c:325
0x58c9e5 prune_lambda_captures
	/tmp/root/spack-stage/spack-stage-gcc-8.1.0-fde6qts62r4baorvcrpkjxc7c6vzmn3w/spack-src/gcc/cp/lambda.c:1489
0x58c9e5 finish_lambda_function(tree_node*)
	/tmp/root/spack-stage/spack-stage-gcc-8.1.0-fde6qts62r4baorvcrpkjxc7c6vzmn3w/spack-src/gcc/cp/lambda.c:1519
0x681239 cp_parser_lambda_body
	/tmp/root/spack-stage/spack-stage-gcc-8.1.0-fde6qts62r4baorvcrpkjxc7c6vzmn3w/spack-src/gcc/cp/parser.c:10692
0x681239 cp_parser_lambda_expression
	/tmp/root/spack-stage/spack-stage-gcc-8.1.0-fde6qts62r4baorvcrpkjxc7c6vzmn3w/spack-src/gcc/cp/parser.c:10187
0x681239 cp_parser_primary_expression
	/tmp/root/spack-stage/spack-stage-gcc-8.1.0-fde6qts62r4baorvcrpkjxc7c6vzmn3w/spack-src/gcc/cp/parser.c:5263
0x692059 cp_parser_postfix_expression
	/tmp/root/spack-stage/spack-stage-gcc-8.1.0-fde6qts62r4baorvcrpkjxc7c6vzmn3w/spack-src/gcc/cp/parser.c:7022
0x6929b1 cp_parser_unary_expression
	/tmp/root/spack-stage/spack-stage-gcc-8.1.0-fde6qts62r4baorvcrpkjxc7c6vzmn3w/spack-src/gcc/cp/parser.c:8317
0x674f71 cp_parser_cast_expression
	/tmp/root/spack-stage/spack-stage-gcc-8.1.0-fde6qts62r4baorvcrpkjxc7c6vzmn3w/spack-src/gcc/cp/parser.c:9085
0x67576a cp_parser_binary_expression
	/tmp/root/spack-stage/spack-stage-gcc-8.1.0-fde6qts62r4baorvcrpkjxc7c6vzmn3w/spack-src/gcc/cp/parser.c:9186
0x676e44 cp_parser_assignment_expression
	/tmp/root/spack-stage/spack-stage-gcc-8.1.0-fde6qts62r4baorvcrpkjxc7c6vzmn3w/spack-src/gcc/cp/parser.c:9481
0x678b26 cp_parser_parenthesized_expression_list
	/tmp/root/spack-stage/spack-stage-gcc-8.1.0-fde6qts62r4baorvcrpkjxc7c6vzmn3w/spack-src/gcc/cp/parser.c:7756
0x691bf0 cp_parser_postfix_expression
	/tmp/root/spack-stage/spack-stage-gcc-8.1.0-fde6qts62r4baorvcrpkjxc7c6vzmn3w/spack-src/gcc/cp/parser.c:7096
0x6929b1 cp_parser_unary_expression
	/tmp/root/spack-stage/spack-stage-gcc-8.1.0-fde6qts62r4baorvcrpkjxc7c6vzmn3w/spack-src/gcc/cp/parser.c:8317
0x674f71 cp_parser_cast_expression
	/tmp/root/spack-stage/spack-stage-gcc-8.1.0-fde6qts62r4baorvcrpkjxc7c6vzmn3w/spack-src/gcc/cp/parser.c:9085
0x67576a cp_parser_binary_expression
	/tmp/root/spack-stage/spack-stage-gcc-8.1.0-fde6qts62r4baorvcrpkjxc7c6vzmn3w/spack-src/gcc/cp/parser.c:9186
0x676e44 cp_parser_assignment_expression
	/tmp/root/spack-stage/spack-stage-gcc-8.1.0-fde6qts62r4baorvcrpkjxc7c6vzmn3w/spack-src/gcc/cp/parser.c:9481
0x675f83 cp_parser_constant_expression
	/tmp/root/spack-stage/spack-stage-gcc-8.1.0-fde6qts62r4baorvcrpkjxc7c6vzmn3w/spack-src/gcc/cp/parser.c:9765
0x676dc7 cp_parser_initializer_clause
	/tmp/root/spack-stage/spack-stage-gcc-8.1.0-fde6qts62r4baorvcrpkjxc7c6vzmn3w/spack-src/gcc/cp/parser.c:21942
0x676f6a cp_parser_assignment_expression
	/tmp/root/spack-stage/spack-stage-gcc-8.1.0-fde6qts62r4baorvcrpkjxc7c6vzmn3w/spack-src/gcc/cp/parser.c:9502
```

If this is a compiler bug that others will encounter, perhaps we should add a conflict statement for `%gcc@8.1.0` ?

@mic84 @asalmgren @eschnett

--- a/var/spack/repos/builtin/packages/amrex/package.py
+++ b/var/spack/repos/builtin/packages/amrex/package.py
@@ -91,6 +91,11 @@ class Amrex(CMakePackage, CudaPackage, ROCmPackage):
     conflicts('%apple-clang')
     conflicts('%clang')
 
+    # these versions of gcc have lambda function issues
+    # see https://github.com/spack/spack/issues/22310
+    conflicts('%gcc@8.1.0:8.3.0', when='@21.03')
+    conflicts('%gcc@8.1.0:8.2.0', when='@21.01:21.02')
+
     # Check options compatibility
     conflicts('+sundials', when='~fortran',
               msg='AMReX SUNDIALS support needs AMReX Fortran API (+fortran)')

